### PR TITLE
Added fine-tuning task

### LINKED ISSUE #249: test finetuning with a pre-trained model

### PR DESCRIPTION
**Pull request by KennethEnevoldsen**
- bug fix: batch is now always first
- bug fix: Attention mask is passed as intended
- feature: Added fine-tuning task

Opened #250 and #249 

fixes: #245 
edit: 
fixes #249

### CODE REVIEW

File: psycop/common/sequence_models/aggregators.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        super().__init__()
+
+    def forward(
+        self,
+        last_hidden: torch.Tensor,
+        attention_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        return (last_hidden * attention_mask.unsqueeze(-1)).sum(1) / attention_mask.sum(
+            -1,
+        ).unsqueeze(-1)
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/aggregators.py:33]: I assume this is correct, but have no way of knowing ;-)

File: psycop/common/sequence_models/dataset.py [hunk 2]
```diff
@@ -16,3 +16,16 @@ def __len__(self) -> int:
...
+    def __init__(self, patients: list[Patient], labels: list[int]) -> None:
+        self.patients: list[Patient] = patients
+        self.labels: list[int] = labels
+        assert len(self.patients) == len(self.labels)
+
+    def __len__(self) -> int:
+        return len(self.patients)
+
+    def __getitem__(self, idx: int) -> tuple[Patient, int]:
+        return self.patients[idx], self.labels[idx]
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/dataset.py:31]: I'll rewrite this to use `PredictionTime` if you'd like 👍

File: psycop/common/sequence_models/embedders.py [hunk 3]
```diff
@@ -118,7 +118,7 @@ def forward(
         embeddings = word_embed + segment_embed + age_embed + posi_embeddings
         embeddings = self.LayerNorm(embeddings)
         embeddings = self.dropout(embeddings)
-        return embeddings
+        return {"src": embeddings, "src_key_padding_mask": inputs["is_padding"] == 1}
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/embedders.py:121]: Dataclass?

File: psycop/common/sequence_models/tasks.py [hunk 4]
```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
...
+        self.aggregation_module = aggregation_module
+
+        self.d_model: int = embedding_module.d_model
+
+        self.is_binary = num_classes == 1
+        self.num_classes = num_classes
+        if self.is_binary:
+            self.loss = nn.BCEWithLogitsLoss()
+        else:
+            self.loss = nn.CrossEntropyLoss()
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tasks.py:186]: Untested functionality for multiclass?

File: psycop/common/sequence_models/tasks.py [hunk 5]
```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
...
+        self.metrics = self.create_metrics(num_classes)
+
+    @staticmethod
+    def create_metrics(num_classes: int) -> dict[str, Metric]:
+        is_binary = num_classes == 1
+        if is_binary:
+            return {"AUROC": BinaryAUROC()}
+        return {"AUROC (macro)": MulticlassAUROC(num_classes=num_classes)}
+
+    def training_step(self, batch: BatchWithLabels, batch_idx: int) -> torch.Tensor:  # type: ignore # noqa: ARG002
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tasks.py:199]: `# type: ignore ?`

File: psycop/common/sequence_models/tasks.py [hunk 6]
```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
...
+        output = self.forward(batch[0], batch[1])
+        loss = output["loss"]
+        self.log_step("Training", output)
+        return loss
+
+    def validation_step(self, batch: BatchWithLabels, batch_idx: int) -> torch.Tensor:  # type: ignore  # noqa: ARG002
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tasks.py:205]: Should we turn BatchWithLabels into a dataclass?

File: psycop/common/sequence_models/tasks.py [hunk 7]
```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
...
+        output = self.forward(inputs=batch[0], labels=batch[1])
+        self.log_step("Validation", output)
+        return output["loss"]
+
+    def log_step(
+        self,
+        mode: Literal["Validation", "Training"],
+        output: dict[str, torch.Tensor],
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tasks.py:213]: Can we give this a more descriptive name? Alternatively a more descriptive type :-)

File: psycop/common/sequence_models/tasks.py [hunk 8]
```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
...
+    ) -> None:
+        """
+        Logs the metrics for the given mode.
+        """
+        for metric_name, metric in output.items():
+            self.log(f"{mode} {metric_name}", metric)
+
+    def forward(  # type: ignore
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tasks.py:221]:
```suggestion
    def forward(  # type: ignore
```

Why is the type: ignore needed here? :-)

File: psycop/common/sequence_models/tasks.py [hunk 9]
```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
...
+
+        # Aggregate the sequence
+        is_padding = embedded_patients["src_key_padding_mask"]
+        aggregated_patients = self.aggregation_module(
+            encoded_patients,
+            attention_mask=~is_padding,
+        )
+
+        # Classification head
+        logits = self.classification_head(aggregated_patients)
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tasks.py:237]: Use clf-token to mirror their behaviour?

File: psycop/common/sequence_models/tasks.py [hunk 10]
```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
...
+        self.is_binary = num_classes == 1
+        self.num_classes = num_classes
+        if self.is_binary:
+            self.loss = nn.BCEWithLogitsLoss()
+        else:
+            self.loss = nn.CrossEntropyLoss()
+
+        self.classification_head = nn.Linear(self.d_model, num_classes)
+
+        self.metrics = self.create_metrics(num_classes)
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tasks.py:190]: self.metric_fns?

File: psycop/common/sequence_models/tasks.py [hunk 11]
```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
...
+
+    def collate_fn(
+        self,
+        patients_with_labels: list[tuple[Patient, int]],
+    ) -> tuple[dict[str, torch.Tensor], torch.Tensor]:
+        """
+        Takes a list of patients and returns a dictionary of padded sequence ids.
+        """
+        patients, outcomes = list(zip(*patients_with_labels))  # type: ignore
+        patients: list[Patient] = list(patients)
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tasks.py:272]: We probably want to rewrite using PredictionTimes.

File: psycop/common/sequence_models/tests/test_encoder_for_clf.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    encoder = nn.TransformerEncoder(encoder_layer, num_layers=2)
+    return encoder
+
+
+@pytest.fixture()
+def aggregation_module() -> AveragePooler:
+    """
+    A mean pooling module
+    """
+    return AveragePooler()
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tests/test_encoder_for_clf.py:49]: Love to chat about the advantage of using a fixture here 👍

File: psycop/common/sequence_models/tests/test_encoder_for_clf.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        num_classes=1,
+        optimizer_kwargs={"lr": 1e-3},
+        lr_scheduler_kwargs={"num_warmup_steps": 2, "num_training_steps": 10},
+    )
+
+    dataloader = DataLoader(
+        patient_dataset_with_labels,
+        batch_size=32,
+        shuffle=True,
+        collate_fn=clf.collate_fn,  # type: ignore
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tests/test_encoder_for_clf.py:71]: ?

File: psycop/common/sequence_models/tests/test_encoder_for_clf.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    )
+
+    for input_ids, masked_labels in dataloader:
+        output = clf(input_ids, masked_labels)
+        loss = output["loss"]
+        loss.backward()  # ensure that the backward pass works
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tests/test_encoder_for_clf.py:108]: Awesome work here, Kenneth! The tests are becoming nicely brief.

Perhaps we want to split this into an 'integration tests' module over time, so we can run unit and integration tests separately for faster feedback. But not essential that it's right now.

File: requirements.txt [hunk 15]
```diff
@@ -26,6 +26,8 @@ tabulate==0.8.10
 timeseriesflattener==1.6.0
 sentence-transformers==2.2.2
 torch==2.0.1
+lightning==2.0.9
+torchmetrics==1.1.2
```
Comment (MartinBernstorff) [inline requirements.txt:30]: Trying ot keep alphabetical or grouped?

**Review by MartinBernstorff — CHANGES_REQUESTED** [review]

File: psycop/common/sequence_models/tests/test_encoder_for_clf.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def test_pretrain_from_checkpoint(
```
Comment (MartinBernstorff) [inline psycop/common/sequence_models/tests/test_encoder_for_clf.py:80]: ? Merge with pretrain test?

**Review by MartinBernstorff — APPROVED** [review]

File: psycop/common/sequence_models/tests/test_encoder_for_clf.py (hunk 16 above)
Comment (KennethEnevoldsen) [reply psycop/common/sequence_models/tests/test_encoder_for_clf.py:80]: Will keep it as is for now. It is nice to know when we can't reload an existing checkpoint anymore, but have added a docstring for clarity.

File: psycop/common/sequence_models/tasks.py [hunk 17]
```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
...
+
+    def log_step(
+        self,
+        mode: Literal["Validation", "Training"],
+        output: dict[str, torch.Tensor],
+    ) -> None:
+        """
+        Logs the metrics for the given mode.
+        """
+        for metric_name, metric in output.items():
```
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply psycop/common/sequence_models/tasks.py:218]: I actually think it is fine (we just get a distribution vector, can be useful for debugging)

**Comment by MartinBernstorff** [comment]
@KennethEnevoldsen We probably want the "resume-from-checkpoint" test to document how you re-generate the checkpoint in case of changes. Would you take a look at that? :-)

File: psycop/common/sequence_models/dataset.py (hunk 2 above)
Comment (KennethEnevoldsen) [reply psycop/common/sequence_models/dataset.py:31]: moved to issue #263

File: psycop/common/sequence_models/tasks.py (hunk 11 above)
Comment (KennethEnevoldsen) [reply psycop/common/sequence_models/tasks.py:307]: moved to issue #263
